### ISSUE TITLE
Uncouple glyph storage from text processing

### ISSUE THREAD

**Issue by djeedai** [bevyengine/bevy#16901]
## What problem does this solve or what need does it fill?

I have a crate doing text rendering in a different way than Bevy does. Originally, I was reusing the Bevy text pipeline, it was really great. Later some changes forced me to make a copy of the entire `TextPipeline` and related classes, which added a lot of maintenance overhead, just because the way Bevy stored glyphs (one texture per font per size) was in my opinion quite wasteful, and I wanted a different glyph storage. Now that Bevy switched from ab to cosmic, I find myself once again copy-pasting and very lightly modifying the `TextPipeline` and other classes, just because of that storage decision. This is a lot of wasted time, while there's no real good reason other than lack of previous need why the glyph storage solution is intertwined with the text processing of Cosmic.

## What solution would you like?

Add an interface to decouple where/how glyphs as stored in textures, from the rest of the text processing (collecting glyphs, rasterizing them, layouting multi-line text, _etc._).

## What alternative(s) have you considered?

What I'm doing currently, which is copy/pasting `bevy_text` and doing some light modifications. This however invariably lead to subtle issues, which requires days of debugging to figure out why glyphs are not properly aligned _etc._. This works entirely duplicate what `bevy_text` authors already did.

## Additional context

I believe we should also really reconsider that pattern of creating one texture atlas per font per size. It's unlikely any app has more than a few fonts in use, and possibly some of them are used for a few words only (title, _etc._) which means we're wasting large atlas textures and force (in the absence of bindless) separate draw calls for no good reason, as there's no technical limitation I know of to packing different font families and sizes into the same atlas.

**Comment by ickshonpe** [comment]
Yeah this has been bothering me forever too, there are always critical bugs etc to fix though so I've never made a pass at it.

**Comment by rparrett** [comment]
I know there is some interest in alternative text rendering, including not using our rasterizer at all. It probably would be nice for folks to be able to continue using our text layout stuff and their own renderer without resorting to copy/pasting everything text related in Bevy.

I am not sure if abstracting away glyph raster **storage** specifically is the best thing to do here. But if it can be done without introducing a bunch of complexity or compromising performance, then I'm not opposed or anything.

> I believe we should also really reconsider that pattern of creating one texture atlas per font per size.
> ...
> we're wasting large atlas textures

I've been playing around with related improvements to try to form an opinion about this.

At the very least, we need to use separate textures for `TextFont`s with `font_smoothing = false` so the underlying texture can have `ImageSampler::nearest()`.

The only reason I can think of for doing separate textures for each font size is to help with sizing the textures to minimize memory usage, but if our atlas textures could grow, this would no longer be a factor.

(side note) Atlas textures seem to currently be `smallest power of two greater than max glyph dimension of the first processed glyph or 512 if that is bigger` which feels pretty janky. That should probably just be based on font size, and possibly made configurable by the user.

I tried implementing font-atlas-growing and it seems like there probably aren't any major blockers to getting that going. The atlas allocator already supports growing the atlas, but `Image` can't currently be grown in a way that doesn't corrupt the texture. That wasn't too difficult to add though. https://github.com/rparrett/bevy/commit/d9d0022d83e0523ae8ee8e7dd8e6b14889b1aca1

If atlases can grow, then a minimum of 512x512 seems really excessive as a starting point -- that's a huge number of glyphs (probably > 1000?) at the default font size.

**Comment by tigregalis** [comment]
https://github.com/bevyengine/bevy/issues/5636 is an issue from the past with some relevant discussion.

A FontAtlas per Font & Size is a reasonable default, since in most cases there's a small predetermined set, and they aren't typically animated.

But I also want to see the individual stages decoupled and be able to hook into these as needed. My comment from the working group on this: https://discord.com/channels/691052431525675048/1248074018612051978/1248404941077217430

> i think modularising this is a valuable goal
> 
> consider a pipeline with offshoots, supporting different rendering needs at different stages along the traditional pipeline, some rendering techniques build on more "primitive" data. it's more of a tree than a pipeline.
> 
> all need:
> loading (loading the fonts)
> font selection (which fonts for a span)
> shaping (which glyphs from a font)
> layout (glyphs placed where)
> (I think these need to be one "module" since they are mostly bidirectionally interdependent)
> 
> we further take curves, but some rendering techniques may use the curves directly
> 
> we further use the curves to rasterise text in different binary/image formats, and some rendering approaches might want to use those rasterised images directly
> 
> we further pack the rasterised text into atlases, and the traditional rendering approach renders these on the gpu

One thing I'm doing (for faster initial load time) in a project (not bevy-related, but is using cosmic-text) is precomputing the glyph atlas for a set of fonts, sizes and glyphs, and embedding the result in the binary, then at runtime (start-up), loading from that embedded data. Essentially this is equivalent to "baking" assets, but currently the TextPipeline system has no real relation to assets.
